internal/dag/builder.go

```diff
@@ -38,53 +32,63 @@ type RouteServiceName struct {
 	Port      int32
 }
 
+// Processor uses a DAG builder to construct part
+// of a DAG.
+type Processor interface {
+	// SetBuilder provides the Processor with a Builder
+	// to use.
+	SetBuilder(builder *Builder)
+
+	// Run executes the processor.
+	Run()
+}
+
 // Builder builds a DAG.
 type Builder struct {
-
 	// Source is the source of Kubernetes objects
 	// from which to build a DAG.
 	Source KubernetesCache
 
-	// DisablePermitInsecure disables the use of the
-	// permitInsecure field in HTTPProxy.
-	DisablePermitInsecure bool
-
-	services map[RouteServiceName]*Service
+	// Processors is the ordered list of Processors to
+	// use to build the DAG.
+	Processors []Processor
 
+	services           map[RouteServiceName]*Service
 	virtualhosts       map[string]*VirtualHost
 	securevirtualhosts map[string]*SecureVirtualHost
-
-	orphaned map[types.NamespacedName]bool
-
-	FallbackCertificate *types.NamespacedName
+	listeners          []*Listener
 
 	StatusWriter
 	logrus.FieldLogger
 }
 
-// Build builds a new DAG.
+// Build builds and returns a new DAG.
```

Comment:
Add a note about processors? Especially we should note the API guarantee that processors are run in order (maybe add a test for that?).